realm/realm-library/src/objectServer/java/io/realm/internal/async/RealmEventStreamAsyncTaskImpl.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal.async;
+
+import java.io.IOException;
+
+import io.realm.internal.Util;
+import io.realm.internal.objectserver.EventStream;
+import io.realm.mongodb.App;
+import io.realm.mongodb.AppException;
+import io.realm.mongodb.ErrorCode;
+import io.realm.mongodb.RealmEventStreamAsyncTask;
+import io.realm.mongodb.mongo.events.BaseChangeEvent;
+
+public class RealmEventStreamAsyncTaskImpl<T> implements RealmEventStreamAsyncTask<T> {
+    private final String name;
+    private final Executor<T> executor;
+    private volatile EventStream<T> eventStream;
+    private volatile boolean isCancelled;
+    private Thread thread;
+
+    public RealmEventStreamAsyncTaskImpl(final String name, final Executor<T> executor) {
+        Util.checkNull(executor, "name");
+        Util.checkNull(executor, "executor");
+
+        this.executor = executor;
+        this.name = name;
+    }
+
+    @Override
+    public synchronized void get(App.Callback<BaseChangeEvent<T>> callback) throws IllegalStateException {
+        Util.checkNull(callback, "callback");
+
+        if (thread != null) {
+            throw new IllegalStateException("Resource already open");
+        } else {
+            thread = new Thread(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        eventStream = executor.run();
+
+                        while (true) {
+                            BaseChangeEvent<T> nextEvent = eventStream.getNextEvent();
+
+                            callback.onResult(App.Result.withResult(nextEvent));
```

Comment:
@cmelchior Do we have a convention about logging and swalling user exceptions? Seen it somewhere, but maybe it is preferable that users gets the attention?

Reply:
I found this one:
RealmLog.error(exception);

What is the log policy, shall we log all errors? For example, shall we log the IOExceptions in `RealmEventStreamAsyncTaskImpl.java` too?

Reply:
Normally we try to surface crashes to users instead of catching them. There is one noticeable, namely callbacks called from JNI and run in Java.

If there is a Java exception inside those, they don't necessarily crash the JNI code, that only happens if it touches `JNIEnv`, and then the stack trace you get can look really strange. So for those cases, we do attempt to catch all Java exceptions and either convert them to something that JNI understands or log them and attempt to stop the event flow as quickly as possible. But is very tricky and I wouldn't be surprised if we forgot to do that in some places.

Reply:
But for the App Request API, we do have a way to roundtrip errors, so in that case we shouldn't log them, since an AppException is thrown as quickly as possible.

Reply:
I wasn't using the exception generated by the watch stream. The error handling has makes more sense now that I have included support for such exceptions.